### ISSUE TITLE
Bug in color space selection

### ISSUE THREAD

**Issue by ToolChainGang** [6by9/raspi_tc358743#1]
I have a client developing his own Toshiba TC358743 interface to RasPi, and have identified what I believe to be a bug in the driver.

I have hardware that shows the bug, a proposed fix, and I can test any changes on AUVidea B101 and B103 boards as well as my client's board.

I'd like to discuss this with someone who knows more about drivers, verify that the proposed change seems correct, and come up with a plan for getting the fix submitted to the RasPi code. (The fix should also probably be back-ported into the linux code.)

Please drop me a note so we can discuss offline.

pwalsh (at) ToolChainGang (dot) com

**Comment by 6by9** [comment]
raspi_tc358743 was only an example application before the V4L2 drivers were completed. Those are now released so this code is only maintained for historical reasons. I've updated the readme to reflect this.

Why the specific need to discuss offline? I can email on Monday, but the main thing with development of these sorts of things is to get as many eyes on problems as possible. I do have the datasheets for the TC358743, however there are several other experts on the linux-media mailing lists.

**Comment by ToolChainGang** [comment]
> Why the specific need to discuss offline?

Basically, because I'm not a linux developer, not a driver developer, 
and generally don't know the ins and outs of the process of making 
changes and submitting them.

Also, was hoping to discuss the technical merits of the change with 
someone who has knowledge of the chip, since I *don't* have the 
datasheet. Is there's a better fix?

 > raspi_tc358743 was only an example application before the V4L2 drivers
 > were completed. Those are now released so this code is only maintained

We're not using V4L, we're using the UV4L system in order to get a 
WebRTC cast from the system. I'd love to get away from that (they want 
some $6000 for a license to remove the 'damn watermark), but there's no 
good substitute for WebRTC.

It sounds like UV4L is using the older driver, and a quick scan of the 
V4L section shows completely different code and might not have the 
colorspace problem we're seeing.

I have an executable that fixes the problem on the fly and can run in 
usermode - just thought it would be better to put it into the base code.

Thanks for getting back to me - it sounds like this is a non-issue.

Pete

**Comment by 6by9** [comment]
> Basically, because I'm not a linux developer, not a driver developer, and generally don't know the ins and outs of the process of making changes and submitting them.

No problem. Anyone can submit an issue providing as much detail as possible, and then anyone involved in the project can analyse and review. We always try to be as friendly as possible, so no need to be concerned that you're not a linux or driver developer. I'll drop you an email anyway.

As you say UV4L is closed source, although I believe it was loosely based on this app. This app in turn took the I2C comms from the Linux kernel driver.

Using V4L2 should be able to remove UV4L from the loop, and there are a number of WebRTC options which should talk direct to V4L2. There may be a couple of tweaks required to use the Pi hardware acceleration for the video codecs.

**Comment by ToolChainGang** [comment]
Per other engineer's request, here's some more information:

My client has designed a demo board using a Toshiba TC358743 HDMI->CSI2 converter chip feeding into a RasPi to cast the video using WebRTC.

The system works with no issues when using an AUVidea B101 or B103 board. When using the designed board, the video looks like image 1 below, note the colors of the standard color bar are wrong.

We postulate this as selecting the wrong color space for the video conversion. Setting register 0x8573 to value 0x81 appears to correct the issue, resulting in image 2 below. We do this using the following code scrap, run from user mode while video is running:

```
    struct cmds_t Cmd;
    Cmd.addr      = 0x8573;
    Cmd.value     = 0x81;
    Cmd.num_bytes = 2;
    printf("Fixing video output!\n");

    write_regs(i2c_fd,&Cmd,1);

    close(i2c_fd);
    exit(0);

```
Note that I do not have access to the Toshiba datasheet, but I have access to a datasheet for a similar Toshiba chip that probably uses the same registers (as matched by the Toshiba regs .h file in the source code). Expecting colorspace selection as the issue, I tried combinations until I found one that works.

We're using an OREI XD600 VGA to HDMI converter (with no input) to generate the test pattern. The arrows are from an HTML overlay on the web page and can be ignored.

Our Toshiba chip is on a daughterboard that can be removed and replaced by the AUVidea boards to test that things work while keeping everything else the same. Our daughterboard is nothing more than a carrier for the Toshiba chip, containing that chip, an oscillator, and some support passives. Close examination of the AUVidea board shows that it also appears to be "just a carrier" board.

![beforefix](https://user-images.githubusercontent.com/11825951/46968700-ebc95900-d081-11e8-9af1-02c537bb09e5.jpg)


![afterfix](https://user-images.githubusercontent.com/11825951/46968750-0c91ae80-d082-11e8-8103-31b492630e2a.jpg)
